components/SearchSets.tsx

```diff
@@ -1,34 +1,40 @@
 import React, { useEffect, useContext } from 'react';
 import useInput from '../hooks/useInput';
-import Context from '../context';
-import { useRouter } from 'next/router';
+import context from '../context';
+import { useRouter, NextRouter } from 'next/router';
+import { Context } from 'typings';
 
 function SearchSets() {
-    const router = useRouter();
+    const router: NextRouter = useRouter();
     const [searchValue, searchInput, setSearchInputValue] = useInput({
-        defaultValue: '',
         placeholder: 'E.g. Carbon',
     });
-    const { setGlobalState } = useContext(Context);
+    const { state, dispatch } = useContext<Context>(context);
 
     // TODO: this supported the use of search?= query in URL...
-    // useEffect(() => {
-    //     const searchQuery = router.query.search;
-    //     if (searchQuery !== undefined) {
-    //         setSearchInputValue(searchQuery);
-    //         setGlobalState({ searchQuery });
-    //     }
-    // }, [router.query.search]);
+    useEffect(() => {
+        const searchQuery = router.query.search;
+        if (searchQuery !== undefined) {
+            setSearchInputValue(searchQuery);
+            dispatch({
+                type: 'set',
+                payload: { searchQuery: searchValue },
+            });
+        }
+    }, [router.query.search]);
 
     useEffect(() => {
         let timeout: any;
         clearTimeout(timeout);
 
         timeout = setTimeout(() => {
             if (searchValue !== '' || searchValue !== undefined) {
-                setGlobalState({ searchQuery: searchValue });
+                dispatch({
+                    type: 'set',
+                    payload: { searchQuery: searchValue },
+                });
             }
-        }, 200);
+        }, 300);
```

Comment:
One thing I might be cautious about is using timers in hooks, Dan Abramov has an article about it but there are some gotchas that I dont think are completely handled. Look like youre trying to debounce the search input which is great, but this might not work as expected. Its kinda gnarly, but something like this might help: 
https://gist.github.com/grantglidewell/9ebb44eac6143b9bf0f261e6dba9b1a0
that is, if youre trying to debounce this and run the function after the timeout. otherwise youre probably running this function for each keystroke, just delayed.

Reply:
Yeay it's because of the debounce.. I did some research on this. It should not run the function because it should clear the timeout. 
I see that your useDebounce also uses timeout, mind explaining me the difference just for my learnings? 
Feel free to implement this useDebounce hook to use this! I'd be happy if you do if it increases performance!

Reply:
I merge it for now. So we can continue on this new way of handling state.